packages/playground/blueprints/src/lib/resources.ts

```diff
@@ -159,7 +159,15 @@ export class VFSResource extends Resource {
 			this.resource.path
 		);
 		this.progress?.set(100);
-		return new File([buffer], this.name);
+		const file = new File([buffer], this.name);
+
+		// Workaround for File class in Node.js lacking arrayBuffer() method
```

Comment:
Oooh nice spot!

Reply:
As with any "workaround" hack I'm not too happy with it, but it's the only way I could get `writeFile` to work with a `File` instance on the server side (and therefore pass the new tests for install plugin/theme).

Since this line only patches `VFSResource`, future tests for blueprint steps may run into the same issue if they use any other resource type's resolved `File`.

I also wonder if I should file an issue with the Node.js repo about how `File` is missing an `arrayBuffer()` method.

Reply:
> I also wonder if I should file an issue with the Node.js repo about how File is missing an arrayBuffer() method.

@eliot-akira I found the original Node.js PR where the File API was introduced: https://github.com/nodejs/node/pull/45139

It seems like the File class is available in the `buffer` module and, wait for it, it implements the `arrayBuffer()` method!

```
> const File = require('buffer').File;
> const f = new File(['a'], 'world.txt');
> f.arrayBuffer()
Promise { <pending> }
```

Even more interestingly, there is no global `File` object:

```
> File
Uncaught ReferenceError: File is not defined
```

I wonder where the one without the `arrayBuffer` method comes from 🤔

Reply:
The missing `File.arrayBuffer()` was happening only in tests. Could it be that `File` is mocked somehow in that environment?

I found the commit where `File` in Node.js became defined as global name.

- buffer: graduate File from experimental and expose as global https://github.com/nodejs/node/commit/7bc0e6a4e79684a98cb64c61452af00e20f590e7#diff-cbe9ed0255d58a51cf6442116da17f833c9d57c6244dc1efa74760d1c91860e7R38

Oh, it's tagged [v20.2.0](https://github.com/nodejs/node/releases/tag/v20.2.0), [v20.1.0](https://github.com/nodejs/node/releases/tag/v20.1.0), [v20.0.0](https://github.com/nodejs/node/releases/tag/v20.0.0). That's probably it, I'm using Node.js v18.16.0 LTS.

Is there a way that the build tool can shim/patch the use of global `File` to work on server side? If it can be done at build time, it would avoid having to conditionally await dynamic import at run time.

Reply:
I found a way to inspect the `File` instance in the class `VFSResource`.

```ts
import { inspect } from 'util';
...
if (!file.arrayBuffer) {
	console.log(inspect(file, { showHidden: true, depth: 1 }));
	...
}

```

Result:

```
<ref *1> File {
  [Symbol(impl)]: FileImpl {
    _buffer: <Buffer 50 4b 03 04 14 00 02 00 08 00 69 8a be 56 55 7e c5 f4 1c 00 00 00 1f 00 00 00 16 00 00 00 2f 74 65 73 74 2d 70 6c 75 67 69 6e 2f 69 6e 64 65 78 2e 70 ... 120 more bytes>,
    _globalObject: [Window],
    type: '',
    name: 'test-plugin-0.0.1.zip',
    lastModified: 1685459958766,
    [Symbol(wrapper)]: [Circular *1],
    [size]: [Getter]
  },
  name: [Getter],
  lastModified: [Getter],
  [Symbol(Symbol.toStringTag)]: 'File',
  size: [Getter],
  type: [Getter]
}
```

I switched to Node.js latest version `20.2.0`. The `File` object looks the same as above.

The tests for `installPlugin` and `installTheme` are still failing without the `arrayBuffer()` shim. Same error as on Node.js `18.16.0`.

```
TypeError: file.arrayBuffer is not a function
    at Module.fileToUint8Array (/home/eliot/wordpress-playground/packages/playground/blueprints/src/lib/steps/common.ts:25:35)
    at Module.writeFile (/home/eliot/wordpress-playground/packages/playground/blueprints/src/lib/steps/client-methods.ts:292:16)
    at Module.installAsset (/home/eliot/wordpress-playground/packages/playground/blueprints/src/lib/steps/install-asset.ts:39:8)
    at Module.installPlugin (/home/eliot/wordpress-playground/packages/playground/blueprints/src/lib/steps/install-plugin.ts:66:31)
    at run (/home/eliot/wordpress-playground/packages/playground/blueprints/src/lib/compile.ts:191:11)
    at Object.run (/home/eliot/wordpress-playground/packages/playground/blueprints/src/lib/compile.ts:95:21)
    at Module.runBlueprintSteps (/home/eliot/wordpress-playground/packages/playground/blueprints/src/lib/compile.ts:261:2)
    at /home/eliot/wordpress-playground/packages/playground/blueprints/src/lib/steps/install-plugin.spec.ts:44:3
    at runTest (file:///home/eliot/wordpress-playground/node_modules/@vitest/runner/dist/index.js:491:11)
    at runSuite (file:///home/eliot/wordpress-playground/node_modules/@vitest/runner/dist/index.js:594:15)
```

Interestingly, in the Node.js REPL:

```
$ node
Welcome to Node.js v20.2.0.
Type ".help" for more information.
> File
[class File extends Blob]
> File.prototype.arrayBuffer
[Function: arrayBuffer]
```

So it looks like `File` in the tests is a mocked object.

Oh I found it. In the Vite config file for the `blueprints` package, under the `test` property it says `environment: jsdom`.

Searching the issue tracker for JSDOM results in:

- jsdom/jsdom#2555

There's the missing method `arrayBuffer`.

Can we shim JSDOM's global `File` in a test "prelude" file that runs only before tests?

Reply:
I tried to shim JSDOM's global `File` in the tests themselves, in Jest's setup method `beforeEach`, but couldn't get it to work. `File.arrayBuffer` still missing in `VFSResource.resolve()`.

Reply:
Clarified in the comment above the workaround that this is specific to JSDOM. [81bb68f](https://github.com/WordPress/wordpress-playground/pull/427/commits/81bb68f6b8cb4585e571363c43b07819f7c75e63)

Maybe the shim can be removed in the production build for browser/server, since it's only necessary during tests.

Reply:
There's a possibility of moving the polyfill to the front of the app initialization and therefore removing the need to turn every call async.

Do we have an easy entry point? If it's not already async we can probably make it async easier than `File` and then that gives us a convenient place to do further bootstrapping in the future.

```js
const main = async () => {
	await bootstrap();
}

const boostrap = async () => {
	await polyfillFile();
}
```

Reply:
I'm finding that it's tricky to polyfill JSDOM's global `File` to have an `arrayBuffer()` method. I think it's the reason why they haven't added it yet:

> I was going to work on a PR for Blob.arrayBuffer() but looks like it will be a no-go for now as the wpt relies on TextEncoder which is not supported at this time ref: https://github.com/jsdom/jsdom/issues/2524

(From: https://github.com/jsdom/jsdom/issues/2555#issuecomment-1266575509)

My workaround patch in `VFResource` was simpler to add, because there's a variable `buffer` in the local scope. But a more generic `File.prototype.arrayBuffer()` would need to get the buffer from somewhere else, but as far as I can see, the property `Blob.buffer` is not public (or doesn't exist in JSDOM's implementation).